include/llama.h

```diff
@@ -842,29 +857,109 @@ extern "C" {
     // Otherwise, llama_batch.token will be allocated to store n_tokens llama_token
     // The rest of the llama_batch members are allocated with size n_tokens
     // All members are left uninitialized
-    LLAMA_API struct llama_batch llama_batch_init(
+    DEPRECATED(LLAMA_API struct llama_batch llama_batch_init(
+                    int32_t n_tokens,
+                    int32_t embd,
+                    int32_t n_seq_max), "use llama_batch_ext API instead");
+
+    // Frees a batch of tokens allocated with llama_batch_init()
+    DEPRECATED(LLAMA_API void llama_batch_free(struct llama_batch batch),
+            "use llama_batch_ext API instead");
+
+    // Allocates a batch of tokens on the heap that can hold a maximum of n_tokens
+    // Each token can be assigned up to n_seq_max sequence ids
+    // The batch has to be freed with llama_batch_ext_free()
+    LLAMA_API struct llama_batch_ext * llama_batch_ext_init(
             int32_t n_tokens,
-            int32_t embd,
             int32_t n_seq_max);
 
-    // Frees a batch of tokens allocated with llama_batch_init()
-    LLAMA_API void llama_batch_free(struct llama_batch batch);
+    // Same with llama_batch_init, but initializes the batch with the provided text tokens
+    // First token will be at position pos0
+    // The sequence ID will be fixed to seq_id
+    // The batch has to be freed with llama_batch_ext_free()
+    LLAMA_API struct llama_batch_ext * llama_batch_ext_init_from_text(
+            llama_token * tokens,
+                int32_t   n_tokens,
+                int32_t   pos0,
+                int32_t   seq_id);
+
+    // Same with llama_batch_init, but initializes the batch with the provided raw embeddings
+    // First token will be at position pos0
+    // The sequence ID will be fixed to seq_id
+    // The batch has to be freed with llama_batch_ext_free()
+    LLAMA_API struct llama_batch_ext * llama_batch_ext_init_from_embd(
+              float * embd,
+            size_t    n_embd,
+            int32_t   pos0,
+            int32_t   seq_id);
+
+    // Get the number of tokens in the batch
+    LLAMA_API int32_t llama_batch_ext_get_n_tokens(const struct llama_batch_ext * batch);
+
+    LLAMA_API struct llama_batch_ext_token_info llama_batch_ext_get_token_info(
+        struct llama_batch_ext * batch,
+                       int32_t   i);
+
+    // Add text tokens to the batch
+    // Return values:
+    //  0 : success
+    // -1 : not enough space in the batch
+    // -2 : embd is already set, cannot add text tokens
+    LLAMA_API int32_t llama_batch_ext_add_text_token(
```

Comment:
```suggestion
    LLAMA_API int32_t llama_batch_ext_add_text(
```

For consistency with `llama_batch_ext_init_from_text`. Othewise it would have to be `llama_batch_ext_init_from_text_tokens`.